src/main/java/org/apache/mesos/hdfs/Scheduler.java

```diff
@@ -73,18 +81,178 @@ public void registered(SchedulerDriver driver, FrameworkID frameworkId, MasterIn
     try {
       persistentState.setFrameworkId(frameworkId);
     } catch (InterruptedException | ExecutionException e) {
+      log.error("Error setting framework id in persistent state", e);
       throw new RuntimeException(e);
     }
     log.info("Registered framework frameworkId=" + frameworkId.getValue());
+    //reconcile tasks upon registration
+    reconcileTasks(driver);
   }
 
   @Override
   public void reregistered(SchedulerDriver driver, MasterInfo masterInfo) {
-    log.info("Reregistered framework.");
+    log.info("Reregistered framework: starting task reconciliation");
+    // reconcile tasks upon reregistration
+    reconcileTasks(driver);
+  }
+
+  @Override
+  public void statusUpdate(SchedulerDriver driver, TaskStatus status) {
+    log.info(String.format(
+        "Received status update for taskId=%s state=%s message='%s' stagingTasks.size=%d",
+        status.getTaskId().getValue(),
+        status.getState().toString(),
+        status.getMessage(),
+        liveState.getStagingTasksSize()));
+
+    if (!isStagingState(status)) {
+      liveState.removeStagingTask(status.getTaskId());
+    }
+
+    if (isTerminalState(status)) {
+      liveState.removeRunningTask(status.getTaskId());
+      persistentState.removeTaskId(status.getTaskId().getValue());
+      if (reconciliationComplete()) {
```

Comment:
This logic was not initially obvious to me, so maybe it's worth a comment. Please correct me if I'm wrong: If we're in the RECONCILING state and reconciliation just completed (timeout expired), then we need to update to whatever state is appropriate. If we're RECONCILING, but the timeout has not expired, then we stay in that state (no correctCurrentPhase). On the other hand, if we're in any other status (except maybe JN) and a task just died, then we need to update our state to possibly recover from the loss; in any of these cases, reconciliationComplete() will also be true. Thus it is unnecessary to check the state explicitly here.